sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala

```diff
@@ -413,7 +413,7 @@ private[execution] final class LongToUnsafeRowMap(val mm: TaskMemoryManager, cap
 
   private def init(): Unit = {
     if (mm != null) {
-      require(capacity < 512000000, "Cannot broadcast more than 512 millions rows")
+      require(capacity <= 512000000, "Cannot broadcast more than 512 millions rows")
```

Comment:
I think the previous condition was true.

https://github.com/apache/spark/blob/79c66894296840cc4a5bf6c8718ecfd2b08bcca8/sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/BroadcastExchangeExec.scala#L80-L83

negated `numRows >= 512000000` should be `capacity < 512000000`

Reply:
Ah, good catch. Sorry @sekikn could we change this one more time, to use `<` and then fix the error message instead?

Reply:
Sure! I'm going to update the PR.